sdk_v2/kubeflow/training/constants/constants.py

```diff
@@ -0,0 +1,128 @@
+# Copyright 2024 The Kubeflow Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+
+# How long to wait in seconds for requests to the Kubernetes API Server.
+DEFAULT_TIMEOUT = 120
+
+# Common constants.
+GROUP = "kubeflow.org"
+VERSION = "v2alpha1"
+API_VERSION = f"{GROUP}/{VERSION}"
+
+# The default Kubernetes namespace.
+DEFAULT_NAMESPACE = "default"
+
+# The Kind name for the ClusterTrainingRuntime.
+CLUSTER_TRAINING_RUNTIME_KIND = "ClusterTrainingRuntime"
+
+# The plural for the ClusterTrainingRuntime.
+CLUSTER_TRAINING_RUNTIME_PLURAL = "clustertrainingruntimes"
+
+# The label key to identify training phase where TrainingRuntime should be used.
+# For example, runtime for the pre-training or post-training.
+PHASE_KEY = "training.kubeflow.org/phase"
+
+# The value indicates that runtime can be used for the model pre-training.
+PHASE_PRE_TRAINING = "pre-training"
+
+# The value indicates that runtime can be used for the model post-training.
+PHASE_POST_TRAINING = "post-training"
+
+# The label key to identify the device (e.g. GPU, TPU) that is used for training.
+# TODO: Potentially, we should get this data from the Node selectors.
+DEVICE_KEY = "training.kubeflow.org/device"
```

Comment:
Here, what exactly does the "device" mean? 
Computing Device? Persistent Storage device? Ephemeral Storage device like memory and extended persistent memory? NIC device?

Reply:
In that case, should we name it just an accelerator?
The device sounds weird since you seem to aim support only accelerator including FPGA, and you do not have any plan to support any other special devices.

Reply:
It's a good point. However, I was just exploring how make it more close to ML frameworks, and they call their API as "device".
- PyTorch: https://pytorch.org/docs/stable/generated/torch.cuda.device.html
- MLX: https://ml-explore.github.io/mlx/build/html/python/_autosummary/mlx.core.set_default_device.html#mlx.core.set_default_device

Thus, I was thinking that device is closer to data scientists expectation when they want to assign their ML code to the specific device. WDYT @tenzen-y ?

Reply:
For PyTorch, it seems that the device indicates only GPU since the lib name is "cuda.device".
For MLX, I'm not sure which devices can be specified because I can not find any documentation for supported devices at first glance.

As a result, I still would recommend using the accelerator.

Reply:
>For MLX, I'm not sure which devices can be specified because I can not find any documentation for supported devices at first glance

In MLX you can specify CPU or GPU (e.g. Metal). For example: `mx.set_default_device(mx.cpu) or mx.set_default_device(mx.gpu)`: https://github.com/ml-explore/mlx-examples/blob/main/mnist/main.py#L103

On EKS and GKE the label that indicates GPU type on the node is named accelerator `k8s.amazonaws.com/accelerator`: https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/cloudprovider/aws/README.md#special-note-on-gpu-instances

However, how we can deal with CPU-based runtimes, e.g. we don't call CPU an "accelerator", isn't ?
Or maybe in the future we can support other devices/accelerators.  


@tenzen-y Can we create a tracking issue to discuss this further ?

Reply:
> However, how we can deal with CPU-based runtimes, e.g. we don't call CPU an "accelerator", isn't ?
Or maybe in the future we can support other devices/accelerators.

IIUC, this device field corresponds to the Kubernetes scheduling directives like nodeSelector. So, I'm not sure the reason why we need to provide the knob for CPU workloads.

Reply:
@tenzen-y The main goal of this device and device count parameter is to show Data Scientists which hardware resources they can utilize while using the existing Runtimes. With that, they can appropriately configure their training code to use those devices. For example, using `nccl` backend in PyTorch if GPUs are available or using `bfloat16` if GPU supports it.

As you can see, the device and device count is part of Runtime and Component classes that Data Scientists can get: https://github.com/kubeflow/training-operator/blob/e9c22b2b477f4beea5c96006f6997e8212f4ee4c/sdk_v2/kubeflow/training/types/types.py#L24-L39 

Does it make sense ?